src/rpc/mining.cpp

```diff
@@ -948,6 +964,107 @@ static UniValue estimaterawfee(const JSONRPCRequest& request)
+        }
+    }.Check(request);
+
+    if (!Params().GetConsensus().signet_blocks) {
+        throw std::runtime_error("getnewblockhex can only be used with signet networks");
+    }
+
+    CScript coinbase_script;
+    auto str = request.params[0].get_str();
+    if (IsValidDestinationString(str)) {
+        auto dest = DecodeDestination(str);
+        coinbase_script = GetScriptForDestination(dest);
+    } else {
+        util::insert(coinbase_script, ParseHexV(str, "coinbase_script"));
+    }
+
+    auto block = BlockAssembler(Params()).CreateNewBlock(coinbase_script)->block;
+    unsigned int extra_nonce = 0;
+    // we bump stuff to bop stuff, or merkle root will be 0, etc etc etc etc
+    {
+        LOCK(cs_main);
+        IncrementExtraNonce(&block, ::ChainActive().Tip(), extra_nonce);
+    }
+
+    CDataStream ssBlock(SER_NETWORK, PROTOCOL_VERSION);
+    ssBlock << block;
+
+    return HexStr(ssBlock.begin(), ssBlock.end());
+}
+
+static UniValue grindblock(const JSONRPCRequest& request)
+{
+    RPCHelpMan{"grindblock",
+        "\nGrind the given signet block to find valid proof of work\n"
+        "May fail if it reaches maxtries attempts.\n",
+        {
+            {"blockhex", RPCArg::Type::STR, RPCArg::Optional::NO, "The block data, or a script challenge, for new Signet networks"},
+            {"maxtries", RPCArg::Type::NUM, /* default */ "1000000", "How many iterations to try."},
+        },
+        {
+            RPCResult{"for block data",
+                "blockhash     (hex) resulting block hash, or null if none was found\n"
+            },
+            RPCResult{"for script challenge",
+                "nonce         (number) resulting nonce value which satisfies the proof of work requirement"
+            },
+        },
+        RPCExamples{
+            "\nGrind a block with hex $blockhex\n"
+            + HelpExampleCli("grindblock", "$blockhex")
+            + "\nCreate a new signet network with challenge 512103ad5e0edad18cb1f0fc0d28a3d4f1f3e445640337489abb10404f2d1e086be43051ae\n"
+            + HelpExampleCli("grindblock", "512103ad5e0edad18cb1f0fc0d28a3d4f1f3e445640337489abb10404f2d1e086be43051ae")
+        },
+    }.Check(request);
+
+    bool grinding_nonce = false;
+    CBlock block;
+    if (DecodeHexBlk(block, request.params[0].get_str())) {
+        if (!Params().GetConsensus().signet_blocks) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Grinding blocks is only possible in a signet network; if you want to make a new Signet network, provide the challenge instead");
+        }
+
+        std::vector<uint8_t> signet_commitment;
+        if (!block.GetWitnessCommitmentSection(SIGNET_HEADER, signet_commitment)) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Block has no signet commitment; please sign it first");
+        }
+    } else {
+        auto bin = ParseHexV(request.params[0], "challenge");
+        block = CreateSignetGenesisBlock(CScript(bin.begin(), bin.end()));
```

Comment:
Why do we need to mine the genesis block, isn't it easier to just not validate the genesis block beyond its hash defined in chainparams?

Reply:
https://github.com/bitcoin/bitcoin/pull/16630#issuecomment-521988316

Reply:
With https://github.com/bitcoin/bitcoin/pull/17037 we wouldn't need to grind the genesis block, and it is smaller than #8994 .